File: backend/app/schemas/world_sim.py
"""World Simulator schemas (Sprint 9, docs/02 §11.6).

Response schemas mirror the dict shapes produced by
`WorldSimulatorService`; request schemas cover the god-tool endpoints.
"""

from pydantic import BaseModel, Field


class WorldRoadRead(BaseModel):
    from_id: str
    to_id: str
    built_day: int


class WorldSettlementRead(BaseModel):
    id: str
    name: str
    x: int
    y: int
    population: int
    food: int
    level: int
    experience: int
    skill_level: int
    status: str  # "active" | "abandoned"
    terrain: str
    founded_day: int
    destroyed_day: int | None = None
    parent_id: str | None = None
    farmers: int
    builders: int
    merchants: int
    explorers: int


class WorldSettlementDetailRead(WorldSettlementRead):
    roads: list[WorldRoadRead] = []


class WorldEventRead(BaseModel):
    id: str
    day: int
    event_type: str  # discovery | trade | conflict | natural | social
    title: str
    narrative: str
    severity: int = 1  # 1-10
    affected_settlements: list[str] = []


class WorldStatsRead(BaseModel):
    settlements: int
    active: int
    abandoned: int
    population: int
    roads: int
    events: int


class WorldSimStateRead(BaseModel):
    day_number: int
    time_scale: int
    seed: int
    updated_at: str
    settlements: list[WorldSettlementRead]
    roads: list[WorldRoadRead]
    recent_events: list[WorldEventRead]
    stats: WorldStatsRead


class WorldTickRequest(BaseModel):
    days: int = Field(1, ge=1, le=365)


class WorldTickResponse(BaseModel):
    days_advanced: int
    day_number: int


class WorldResetRequest(BaseModel):
    seed: int | None = None


class WorldAccelerateRequest(BaseModel):
    time_scale: int = Field(1, ge=1, le=10)


class WorldDisasterRequest(BaseModel):
    settlement_id: str
    disaster_type: str


class WorldDisasterResponse(BaseModel):
    settlement_id: str
    disaster_type: str
    applied: bool

File: frontend/src/api/world_sim.ts
import { api } from "./client";

export interface WorldSettlement {
  id: string;
  name: string;
  x: number;
  y: number;
  population: number;
  food: number;
  level: number;
  experience: number;
  skill_level: number;
  status: "active" | "abandoned";
  terrain: string;
  founded_day: number;
  destroyed_day: number | null;
  parent_id: string | null;
  farmers: number;
  builders: number;
  merchants: number;
  explorers: number;
}

export interface WorldRoad {
  from_id: string;
  to_id: string;
  built_day: number;
}

export interface WorldEvent {
  id: string;
  day: number;
  event_type: string;
  title: string;
  narrative: string;
  severity: number;
  affected_settlements: string[];
}

export interface WorldStats {
  settlements: number;
  active: number;
  abandoned: number;
  population: number;
  roads: number;
  events: number;
}

export interface WorldState {
  day_number: number;
  time_scale: number;
  seed: number;
  updated_at: string;
  settlements: WorldSettlement[];
  roads: WorldRoad[];
  recent_events: WorldEvent[];
  stats: WorldStats;
}

export interface WorldTickResult {
  days_advanced: number;
  day_number: number;
}

export async function getWorldState(): Promise<WorldState> {
  const { data } = await api.get<WorldState>("/v1/world-sim/state");
  return data;
}

export async function getWorldSettlement(
  settlementId: string,
): Promise<WorldSettlement & { roads: WorldRoad[] }> {
  const { data } = await api.get<WorldSettlement & { roads: WorldRoad[] }>(
    `/v1/world-sim/settlements/${settlementId}`,
  );
  return data;
}

export async function tickWorld(days: number): Promise<WorldTickResult> {
  const { data } = await api.post<WorldTickResult>("/v1/world-sim/tick", {
    days,
  });
  return data;
}

export async function resetWorld(
  seed?: number,
): Promise<{ status: string; seed: number }> {
  const { data } = await api.post<{ status: string; seed: number }>(
    "/v1/world-sim/reset",
    { seed: seed ?? null },
  );
  return data;
}

export async function accelerateWorld(
  timeScale: number,
): Promise<{ time_scale: number }> {
  const { data } = await api.post<{ time_scale: number }>(
    "/v1/world-sim/accelerate",
    { time_scale: timeScale },
  );
  return data;
}

export async function triggerWorldDisaster(
  settlementId: string,
  disasterType: string,
): Promise<{ settlement_id: string; disaster_type: string; applied: boolean }> {
  const { data } = await api.post<{
    settlement_id: string;
    disaster_type: string;
    applied: boolean;
  }>("/v1/world-sim/disaster", {
    settlement_id: settlementId,
    disaster_type: disasterType,
  });
  return data;
}
